pep-0592.rst

```diff
@@ -0,0 +1,280 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Typically when adding a new type, we need to upstream that type to the
+typing module and change MyPy [mypy]_, PyCharm [pycharm]_, Pyre [pyre]_,
+pytype [pytype]_, etc...
+This is particularly important when working on open-source code that
+makes use of our new types, seeing as the code would not be immediately
+transportable to other developers’ tools without additional logic (this
+is a limitation of MyPy plugins [mypy-plugins]_, which allow for extending MyPy but
+would require a consumer of new typehints to be using MyPy and have the
+same plugin installed). As a result, there is a high cost to developing
+and trying out new types in a codebase. Ideally, we should be able to
+introduce new types in a manner that allows for graceful degradation
+when clients do not have a custom MyPy plugin, which would lower the
+barrier to development and ensure some degree of backward compatibility.
+
+For example, suppose that we wanted to add support for tagged unions
+[tagged-union]_ to Python. One way to accomplish would be to annotate TypedDict
+[typed-dict]_ in Python such that only one field is allowed to be set::
+
+    Currency = Annotated(
+      TypedDict('Currency', {'dollars': float, 'pounds': float}, total=False),
+      TaggedUnion,
+    )
+
+This is a somewhat cumbersome syntax but it allows us to iterate on this
+proof-of-concept and have people with non-patched IDEs work in a
+codebase with tagged unions. We could easily test this proposal and iron
+out the kinks before trying to upstream tagged union to ``typing``,
+MyPy, etc. Moreover, tools that do not have support for parsing the
+``TaggedUnion`` annotation would still be able able to treat
+``Currency`` as a ``TypedDict``, which is still a close approximation
+(slightly less strict).
+
+Specification
+-------------
+
+Syntax
+~~~~~~
+
+``Annotated`` is parameterized with a type and an arbitrary list of
+Python values that represent the annotations. Here are the specific
+details of the syntax:
+
+-  The first argument to ``Annotated`` must be a valid ``typing`` type
+-  Multiple type annotations are supported (Annotated supports variadic
+   arguments): ``Annotated[int, ValueRange(3, 10), ctype("char")]``
+-  ``Annotated`` must be called with at least two arguments
+   (``Annotated[int]`` is not valid)
+-  The order of the annotations is preserved and matters for equality
+   checks:
+   ``Annotated[int, ValueRange(3, 10), ctype("char")] !=``\ Annotated[int,
+   ``ctype("char"),``\ ValueRange(3, 10)]```\`
+-  Nested ``Annotated`` types are flattened, with metadata ordered
+   starting with the innermost annotation:
+   ``Annotated[Annotated[int, ValueRange(3, 10)], ctype("char")] ==``\ Annotated[int,
+   ValueRange(3, 10), ctype(“char”)]`\`
+-  Duplicated annotations are not removed:
+   ``Annotated[int,``\ ValueRange(3, 10)\ ``] !=``\ Annotated[int,
+   ``ValueRange(3, 10)``, ``ValueRange(3, 10)``]```````\`
```

Comment:
Something is wrong here and above with lots of `` ` `` symbols and lots of `\` symbols.